pkg/typesystem/typesystem.go

```diff
@@ -667,163 +667,205 @@ func (t *TypeSystem) relationInvolvesExclusion(objectType, relation string, visi
 	return false, nil
 }
 
-// hasEntrypoints recursively walks the rewrite definition for the given relation to determine if there is at least
-// one path in the rewrite rule that could relate to at least one concrete object type. If there is no such path that
-// could lead to at least one relationship with some object type, then false is returned along with an error indicating
-// no entrypoints were found. If at least one relationship with a specific object type is found while walking the rewrite,
-// then true is returned along with a nil error.
+type relationDetails struct {
+
+	// hasEntrypoints is true when a particular object+relation relates to at least one assignable object type.
+	// A relation such as `define document#viewer: [document#viewer]` implies that `hasEntrypoints=false`.
+	// A relation such as `define document#viewer: [user, document#viewer]` implies that `hasEntrypoints=true`.
+	hasEntrypoints bool
+
+	// hasLoop is true when a particular object+relation introduces a loop in the model.
+	// hasEntrypoints=true implies that hasLoop=false.
+	// If hasEntrypoints=false, nothing can be said about hasLoop.
+	// Loops are introduced by computed usersets.
+	hasLoop bool
+}
+
+// getRelationDetails returns information about a particular type and relation.
 // This function assumes that all other model validations have run.
-func hasEntrypoints(
+func (t *TypeSystem) getRelationDetails(
+	typeName string,
+	relationName string,
+) (*relationDetails, error) {
+	inputRelation, _ := t.GetRelation(typeName, relationName)
+	// we can ignore the error because the getRelationDetailsRecursive will check for it
+	rewrite := inputRelation.GetRewrite()
+	return getRelationDetailsRecursive(t.GetAllRelations(), typeName, relationName, rewrite, map[string]map[string]*relationDetails{})
+}
+
+func getRelationDetailsRecursive(
 	typedefs map[string]map[string]*openfgav1.Relation,
 	typeName, relationName string,
 	rewrite *openfgav1.Userset,
-	visitedRelations map[string]map[string]struct{},
-) (bool, bool, error) {
+	visitedRelations map[string]map[string]*relationDetails,
+) (*relationDetails, error) {
 	v := maps.Clone(visitedRelations)
 
-	if val, ok := v[typeName]; ok {
-		val[relationName] = struct{}{}
-	} else {
-		v[typeName] = map[string]struct{}{
-			relationName: {},
-		}
+	if _, ok := v[typeName]; !ok {
+		// first time visiting this type
+		v[typeName] = map[string]*relationDetails{}
+	}
+	if _, ok := v[typeName][relationName]; !ok {
+		// first time visiting this relation
+		v[typeName][relationName] = &relationDetails{false, false}
```

Comment:
The map was updated not just to track which usersets were visited, but also their details.